rcl/test/rcl/test_service.cpp

```diff
@@ -420,3 +423,227 @@ TEST_F(CLASSNAME(TestServiceFixture, RMW_IMPLEMENTATION), test_service_fail_name
   EXPECT_EQ(RCL_RET_SERVICE_NAME_INVALID, ret) << rcl_get_error_string().str;
   rcl_reset_error();
 }
+
+// Define dummy comparison operators for rcutils_allocator_t type for use with the Mimick Library
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcutils_allocator_t, ==)
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcutils_allocator_t, <)
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcutils_allocator_t, >)
+MOCKING_UTILS_BOOL_OPERATOR_RETURNS_FALSE(rcutils_allocator_t, !=)
+
+/* Mock init failed tests
+ */
+TEST_F(CLASSNAME(TestServiceFixture, RMW_IMPLEMENTATION), test_fail_ini_mocked) {
+  const rosidl_service_type_support_t * ts = ROSIDL_GET_SRV_TYPE_SUPPORT(
+    test_msgs, srv, BasicTypes);
+  const char * topic = "topic";
+  rcl_service_t service = rcl_get_zero_initialized_service();
+  rcl_service_options_t service_options = rcl_service_get_default_options();
+  service_options.qos.durability = RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL;
+  rcl_ret_t ret = RCL_RET_OK;
+
+  {
+    auto mock = mocking_utils::patch_and_return(
+      "lib:rcl", rcutils_string_map_init, RCUTILS_RET_ERROR);
+    ret = rcl_service_init(&service, this->node_ptr, ts, topic, &service_options);
+    EXPECT_EQ(RCL_RET_ERROR, ret);
+    EXPECT_TRUE(rcl_error_is_set());
+    rcl_reset_error();
+  }
+  {
+    // Mocking this function causes rcl_expand_topic_name to return RCL_RET_ERROR
+    auto mock = mocking_utils::patch_and_return(
+      "lib:rcl", rmw_validate_namespace, RMW_RET_ERROR);
+    ret = rcl_service_init(&service, this->node_ptr, ts, topic, &service_options);
+    EXPECT_EQ(RCL_RET_ERROR, ret);
+    EXPECT_TRUE(rcl_error_is_set());
+    rcl_reset_error();
+  }
+  {
+    auto mock = mocking_utils::inject_on_return(
+      "lib:rcl", rcutils_string_map_fini, RCUTILS_RET_ERROR);
+    ret = rcl_service_init(&service, this->node_ptr, ts, topic, &service_options);
+    EXPECT_EQ(RCL_RET_ERROR, ret);
+  }
+  {
+    // Making rcutils_string_map_init fail the second time causes rcl_remap_service_name
+    // to fail
+    auto mock = mocking_utils::patch(
+      "lib:rcl", rcutils_string_map_init,
+      [base = rcutils_string_map_init](
+        rcutils_string_map_t * string_map, size_t initial_capacity, rcutils_allocator_t allocator)
+      {
+        static int counter = 1;
```

Comment:
What role does counter play here? It doesn't look like it's used.

Reply:
It makes the mocked function to fail except the first time is called. As counter is declared static, the increased value is saved (sort of a bomb mocked `rcutils_string_map_init`)

Reply:
Oh, I see what it's doing. Why not just use a `bool`?

Reply:
Removed d8793a9